zerver/decorator.py

```diff
@@ -681,32 +698,30 @@ def wrapper(func):
         def wrapped_func(request, *args, **kwargs):
             # type: (HttpRequest, *Any, **Any) -> HttpResponse
 
-            # It is really tempting to not even wrap our original function
-            # when settings.RATE_LIMITING is False, but it would make
-            # for awkward unit testing in some situations.
-            if not settings.RATE_LIMITING:
-                return func(request, *args, **kwargs)
-
             if client_is_exempt_from_rate_limiting(request):
                 return func(request, *args, **kwargs)
 
-            try:
-                user = request.user
-            except Exception:
-                # TODO: This logic is not tested, and I'm not sure we are
-                # doing the right thing here.
-                user = None
-
-            if not user:
-                logging.error("Requested rate-limiting on %s but user is not authenticated!" %
-                              func.__name__)
-                return func(request, *args, **kwargs)
-
-            # Rate-limiting data is stored in redis
-            # We also only support rate-limiting authenticated
-            # views right now.
-            # TODO(leo) - implement per-IP non-authed rate limiting
-            rate_limit_user(request, user, domain)
+            if settings.RATE_LIMITING:
+                # User based rate limiting. Works when a user is logged in.
+                try:
+                    user = request.user
+                except Exception:
+                    # Not authed. Ignore and move to IP based rate limiting.
+                    logging.debug("Requested rate-limiting on {} but user is "
+                                  "not authenticated!".format(func.__name__))
+                else:
+                    # This function can raise RateLimited exception which is
+                    # processed by
+                    # zerver.middleware.RateLimitMiddleware.process_exception.
+                    rate_limit_user(request, user, domain)
+
+            if settings.IP_RATE_LIMITING:
+                ip = get_ip(request)
```

Comment:
What does this `ipware` library do exactly? I looked at [its repo/README](https://github.com/un33k/django-ipware), but it's not particularly clear -- it just asserts that it does the right things, whatever those are.

I think getting "what is the real user's IP" completely right isn't possible to do without knowing something about the configuration of the network around the server, in particular the layers of proxies the request may come through. Fortunately, for this rate-limiting purpose, it's OK if in that case we end up not unwrapping *enough* proxies -- that will just mean that more requests are grouped together under the same rate-limit bucket when they shouldn't be, which means the rate-limit will bind tighter. In a big site where that's a problem, the admins can configure things to unwrap source-IP information from the proxies they have.

So I think it's enough if we set this up so that
* in the default configuration, we identify the IP address that the request *came to Zulip* through -- to the `nginx` server that the Zulip installer takes responsibility for setting up
* there's something admins can put in `settings.py`, if they like, to identify the IP address behind any further proxies they have.

It looks like `ipware` might be as reasonable a way to do that as any, but we should make sure we're giving it a configuration we're happy with. That'll take a bit of poking, because they don't really have docs.

Alternatively, we could ship with IP rate-limiting disabled by default, and document that admins need to figure out the proxy-unwrapping if they enable it. Because the code and config for getting the source IP could end up being pretty fiddly, it might make sense to merge a version like that first, and figure out this part afterward.